File: DSA/src/Questions/String/SentenceScreenFitting.java
package Questions.String;
/*
https://leetcode.com/problems/sentence-screen-fitting/description/
 Given a rows x cols screen and a sentence represented as a list of strings, return the number of times the given sentence can be fitted on the screen.

The order of words in the sentence must remain unchanged, and a word cannot be split into two lines. A single space must separate two consecutive words in a line.

 

Example 1:

Input: sentence = ["hello","world"], rows = 2, cols = 8
Output: 1
Explanation:
hello---
world---
The character '-' signifies an empty space on the screen.
Example 2:

Input: sentence = ["a", "bcd", "e"], rows = 3, cols = 6
Output: 2
Explanation:
a-bcd- 
e-a---
bcd-e-
The character '-' signifies an empty space on the screen.
Example 3:

Input: sentence = ["i","had","apple","pie"], rows = 4, cols = 5
Output: 1
Explanation:
i-had
apple
pie-i
had--
The character '-' signifies an empty space on the screen.
 

Constraints:

1 <= sentence.length <= 100
1 <= sentence[i].length <= 10
sentence[i] consists of lowercase English letters.
1 <= rows, cols <= 2 * 104
Time complexity: O(rows * cols)
Space complexity: O(n), where n is the length of the concatenated string formed by the sentence.
 */
public class SentenceScreenFitting {
    public int wordsTyping(String[] sentence, int rows, int cols) {
        // Concatenate all words in the sentence into a single string with spaces
        // between words
        // Adding an extra space at the end to handle edge cases where the last word
        // fits perfectly at the end of a row
        String s = String.join(" ", sentence) + " ";

        // Start keeps track of the current position in the concatenated string
        int start = 0;
        int len = s.length(); // Length of the concatenated string

        // Iterate over each row
        for (int i = 0; i < rows; i++) {
            start += cols; // Move the start pointer by the number of columns
            if (s.charAt(start % len) == ' ') {
                // If we land on a space, increment start to the next character
                start++;
            } else {
                // If we land in the middle of a word, backtrack to the previous space
                while (start > 0 && s.charAt((start - 1) % len) != ' ') {
                    start--;
                }
            }
        }

        // The number of times the sentence fits on the screen is given by start / len
        return start / len;
    }

    public static void main(String[] args) {
        SentenceScreenFitting sentenceScreenFitting = new SentenceScreenFitting();
        String[] sentence = { "i", "had", "apple", "pie" };
        int rows = 4;
        int cols = 5;
        System.out.println(sentenceScreenFitting.wordsTyping(sentence, rows, cols));
    }
}

File: DSA/Python/Questions/String/SentenceScreenFitting.py
# Question: Sentence Screen Fitting
# Link: https://leetcode.com/problems/sentence-screen-fitting/

# Problem Statement:
# Given a rows x cols screen and a sentence represented as a list of strings, 
# find how many times the given sentence can be fitted on the screen.

from typing import List


class SentenceScreenFitting:
    def words_typing_dp(self, sentence: List[str], rows: int, cols: int) -> int:
        """
        Dynamic Programming approach using word index tracking.
        Time Complexity: O(rows * cols), where rows is the number of rows and cols is the number of columns.
        Space Complexity: O(1).
        """
        # Join the sentence into a single string with spaces and a trailing space
        s = " ".join(sentence) + " "
        n = len(s)
        chars_fitted = 0

        # Iterate through each row
        for _ in range(rows):
            # Add the number of columns to chars_fitted
            chars_fitted += cols
            # If we land on a space, proceed to the next character
            if s[chars_fitted % n] == " ":
                chars_fitted += 1
            else:
                # Move back to the last valid space
                while chars_fitted > 0 and s[(chars_fitted - 1) % n] != " ":
                    chars_fitted -= 1

        return chars_fitted // n

    def words_typing_optimized(self, sentence: List[str], rows: int, cols: int) -> int:
        """
        Optimized approach using pre-computation.
        Time Complexity: O(rows + len(sentence)), where len(sentence) is the total sentence length.
        Space Complexity: O(1).
        """
        s = " ".join(sentence) + " "
        n = len(s)
        chars_fitted = 0

        # Iterate through each row
        for _ in range(rows):
            chars_fitted += cols
            # Adjust the position based on the character fit
            if s[chars_fitted % n] == " ":
                chars_fitted += 1
            else:
                while chars_fitted > 0 and s[(chars_fitted - 1) % n] != " ":
                    chars_fitted -= 1

        return chars_fitted // n


def main():
    test_cases = [
        {"sentence": ["hello", "world"], "rows": 2, "cols": 8, "expected": 1},
        {"sentence": ["a", "bcd", "e"], "rows": 3, "cols": 6, "expected": 2},
        {"sentence": ["I", "had", "apple", "pie"], "rows": 4, "cols": 5, "expected": 1},
        {"sentence": ["a"], "rows": 5, "cols": 5, "expected": 15},
    ]

    solution = SentenceScreenFitting()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Sentence: {test['sentence']}")
        print(f"Rows: {test['rows']}, Cols: {test['cols']}")
        result_dp = solution.words_typing_dp(
            test["sentence"], test["rows"], test["cols"]
        )
        print(f"DP Result: {result_dp}")
        result_optimized = solution.words_typing_optimized(
            test["sentence"], test["rows"], test["cols"]
        )
        print(f"Optimized Result: {result_optimized}")
        print(f"Expected: {test['expected']}")
        assert result_dp == test["expected"], f"Test case {i + 1} failed"
        print(f"Test case {i + 1} passed")


if __name__ == "__main__":
    main()
